versioned_docs/version-2021fa/lecture9.md

```diff
@@ -0,0 +1,546 @@
+:::
+
+## Security
+
+Web app security is integral in a world where people have malicious intentions. As such, we advocate for at least these basic security measures.
+
+### 1. Passwords
+
+Do NOT store passwords in plain text in your database!!
+
+If somehow there were a vulnerability in your database, the passwords of all of your users would be directly exposed to hackers.
+Moreover, many people reuse their passwords for some if not all other, so you could also compromise something like someone's banking credentials or emails.
+
+While it is safer to store hashed passwords, it is very easy to do it incorrectly, which is why we teach firebase authentication!
+
+### 2. Input Sanitization
+
+Input sanitization 'cleanses' inputs so that they can not be used in unintended ways.
+
+For example, it may seem reasonable to add every comment on a blog into a database entry like this:
+`“INSERT INTO 'comments' (comment) VALUES '” + user_input + “';”`, but it is actually incredibly unsafe
+if `user_input` were something like `'; DROP TABLE comments;`.
+
+Luckily for us, [React DOM escapes values embedded in JSX before rendering them by default, preventing injection attacks](https://reactjs.org/docs/introducing-jsx.html#jsx-prevents-injection-attacks).
+
+### 3. Don't Rely on Client Side Verification
+
+In general you cannot assume your endpoint will only be used by your frontend.
+
+Recall that we used Postman at the beginning of the semester to test Express endpoints. As such, it is always important to verify your inputs and authenticate your requests!
+
+### 4. Separation of Privileges
+
+Going hand-in-hand with the last point, when a request is made to your backend, make sure that the source of your request has the sufficient permissions to perform whatever action is being asked!
+
+For example, if the OfficeHours app recieved a request to edit the times to someone's office hours, the backend would check that this person is indeed the TA who owns that OH and not a student.
+
+### Let's hack our app!
+
+In the lecture we didn't have time to show you guys how we could hack our application. However, if you are interested you can do so by sending in bad unauthenticated input data.
+
+We fixed the backend by using Firebase Admin's verifyIdToken function to check whether the user had logged in or not.
+
+#### Backend
+
+On the backend, we had the `post` endpoints check the `idtoken` in the request headers.
```

Comment:
Let's just emphasize that the header here is called `authorization`, which is pretty standard